ocrd_models/ocrd_models/ocrd_mets.py

```diff
@@ -445,52 +660,90 @@ def set_physical_page_for_file(self, pageId, ocrd_file, order=None, orderlabel=N
+
     def get_physical_page_for_file(self, ocrd_file):
         """
         Get the physical page ID (``@ID`` of the physical ``mets:structMap`` ``mets:div`` entry)
         corresponding to the ``mets:file`` :py:attr:`ocrd_file`.
         """
-        ret = self._tree.getroot().xpath(
-            '/mets:mets/mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"][./mets:fptr[@FILEID="%s"]]/@ID' %
-            ocrd_file.ID, namespaces=NS)
-        if ret:
+        ret = []
+        if self._cache_flag:
+            for pageId in self._fptr_cache.keys():
+                if ocrd_file.ID in self._fptr_cache[pageId].keys():
+                    ret.append(self._page_cache[pageId].get('ID'))
+        else:
+            ret = self._tree.getroot().xpath(
+                '/mets:mets/mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"][./mets:fptr[@FILEID="%s"]]/@ID' %
+                ocrd_file.ID, namespaces=NS)
+
+        # To get rid of the python's FutureWarning
+        if len(ret):
             return ret[0]
 
     def remove_physical_page(self, ID):
         """
         Delete page (physical ``mets:structMap`` ``mets:div`` entry ``@ID``) :py:attr:`ID`.
         """
-        mets_div = self._tree.getroot().xpath(
-            'mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"][@ID="%s"]' % ID,
-            namespaces=NS)
-        if mets_div:
+        mets_div = None
+        if self._cache_flag:
+            if ID in self._page_cache.keys():
+                mets_div = [self._page_cache[ID]]
+        else:
+            mets_div = self._tree.getroot().xpath(
+                'mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"][@ID="%s"]' % ID,
+                namespaces=NS)
+        if mets_div is not None:
             mets_div[0].getparent().remove(mets_div[0])
+            if self._cache_flag:
+                del self._page_cache[ID]
+                del self._fptr_cache[ID]
 
     def remove_physical_page_fptr(self, fileId):
         """
         Delete all ``mets:fptr[@FILEID = fileId]`` to ``mets:file[@ID == fileId]`` for :py:attr:`fileId` from all ``mets:div`` entries in the physical ``mets:structMap``.
         Returns:
             List of pageIds that mets:fptrs were deleted from
         """
-        mets_fptrs = self._tree.getroot().xpath(
-            'mets:structMap[@TYPE="PHYSICAL"]/mets:div[@TYPE="physSequence"]/mets:div[@TYPE="page"]/mets:fptr[@FILEID="%s"]' % fileId,
-            namespaces=NS)
+
+        # Question: What is the reason to keep a list of mets_fptrs?
+        # Do we have a situation in which the fileId is same for different pageIds ?
+        # From the examples I have seen inside 'assets' that is not the case
+        # and the mets_fptrs list will always contain a single element.
+        # If that's the case then we do not need to iterate 2 loops, just one.
```

Comment:
My idea back then as far as I remember was that probably we don't need the second loop inside the `remove_physical_page_fptr()` when the cache is used. However, I am no longer sure how I thought this could be achieved.

Reply:
So I'll remove that comment, ok?